cupy/lib/_routines_poly.py

```diff
@@ -103,6 +103,65 @@ def polymul(a1, a2):
     return cupy.convolve(a1, a2)
 
 
+def polyval(p, x):
+    """Evaluates a polynomial at specific values.
+
+    Args:
+        p (cupy.ndarray or cupy.poly1d): input polynomial.
+        x (scalar, cupy.ndarray): values at which the polynomial
+        is evaluated.
+
+    Returns:
+        cupy.ndarray or cupy.poly1d: polynomial evaluated at x.
+
+    .. warning::
+
+        This function doesn't currently support poly1d values to evaluate.
+
+    .. seealso:: :func:`numpy.polyval`
+
+    """
+    if isinstance(p, cupy.poly1d):
+        p = p.coeffs
+    if not isinstance(p, cupy.ndarray) or p.ndim == 0:
+        raise TypeError('p can be 1d ndarray or poly1d object only')
+    if p.ndim != 1:
+        # to be consistent with polyarithmetic routines' behavior of
+        # not allowing multidimensional polynomial inputs.
+        raise ValueError('p can be 1d ndarray or poly1d object only')
+        # TODO(Dahlia-Chehata): Support poly1d x
+    if (isinstance(x, cupy.ndarray) and x.ndim <= 1) or numpy.isscalar(x):
+        val = cupy.asarray(x).reshape(-1, 1)
+    else:
+        raise NotImplementedError(
+            'poly1d or non 1d values are not currently supported')
+    out = p[::-1] * cupy.power(val, cupy.arange(p.size))
+    out = out.sum(axis=1)
+    dtype = cupy.result_type(p, val)
+    # For case: when p is of shape (0,) and x is (), output is
+    #  of single valued array to match NumPy's behavior
+    if isinstance(x, cupy.ndarray) and x.ndim == 0 and p.size == 0:
+        return out[0].astype(dtype, copy=False)
+    # to match the shape of NumPy's results
+    if cupy.isscalar(x) or x.ndim == 0:
+        return out.astype(dtype, copy=False).reshape()
+    # to match the dtype of NumPy's results
+    if p.dtype == numpy.complex128 and val.dtype in [
+       numpy.float16, numpy.float32, numpy.complex64]:
+        return out.astype(numpy.complex64, copy=False)
+    # To handle mixed integer and float dtypes combinations for inputs,
+    # output should be cast according to NumPy's promotion rules.
+    if p.dtype.kind in 'c' or (issubclass(
+            val.dtype.type, numpy.integer) and issubclass(
+            p.dtype.type, numpy.floating)):
+        return out.astype(dtype, copy=False)
+    # To handle bool values used in evaluation, casting the output
+    # to polynomial dtype is required to match NumPy's results.
+    if val.dtype.kind in 'b':
+        return out.astype(p.dtype, copy=False)
+    return out.astype(val.dtype, copy=False)
```

Comment:
This implementation feels complicated. `get_kind_score` defined in `cupy.core._kernel` will be useful.
```py
    if cupy.isscalar(x) or x.ndim == 0:
        return out.astype(cupy.result_type(p, val), copy=False).reshape()
    p_kind_score = _kernel.get_kind_score(ord(p.dtype.kind))
    x_kind_score = _kernel.get_kind_score(ord(val.dtype.kind))
    if p_kind_score == x_kind_score:
        return out.astype(val.dtype, copy=False)
    return out.astype(cupy.result_type(p, val), copy=False)
```

Reply:
This may sound stupid :sweat_smile: but how to access/import `get_kind_score` ?

Reply:
Oh, `_kernel.get_kind_score` is a `cdef` function, so we cannot import it only from cython file..., Could you change it to `cpdef`?